hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -17,20 +16,48 @@ case class ShuffledStage(child: TableStage)
 case class Binding(name: String, value: IR)
 
 case class TableStage(
-  broadcastVals: IR,
-  globalsField: String,
+  broadcastVals: Array[(String, IR)],
```

Comment:
It looks like the `broadcastVals` IRs are getting duplicated. For example, in `TableCollect`, both `lowered.toIR` and `lowered.globalsWithContext` add let bindings for the broadcast vals. And `TableMapGlobals` adds a let binding for the old globals, but the old globals remain in the broadcast vals, and will be let bound again.

In general, this design seems to make it too easy to duplicate code. Now that I've had a chance to digest this a bit, I'd be happy to chat more about what you want to achieve with this refactoring. Maybe I can help come up with a design that makes it hard to duplicate code.

Reply:
Yeah, for sure.

I've been pretty careful to not *actually* duplicate IR in the existing code---`globalsWithContext` is only called at a point where `toIR` isn't, and if you look carefully `TableMapGlobals` actually only adds a let binding for the reference (basically just so we don't have to actually substitute in values for the `Ref('global', ...)` part in `newGlobal`, and that would ideally be lifted out during optimization), but I struggled a bit with trying to figure out how to structure broadcast values so you wouldn't need to make this distinction, and I think the bigger problem is it's really easy to accidentally use things that don't necessarily exist in a certain context.